feat: weighted candidate scoring w/ biased random selection

rvagg opened the pull request:
* Removed the failure counting and suspension
* Incorporated failure and success counting, with exponential moving average (EMA) (α=0.5)
* Store connect and manipulate time as milliseconds rather than `Duration` for EMA (α=0.5)
* Accumulate an EMA (α=0.8) of _all_ connect times ("overall_ema" below), higher α means newer times turn the ship slower
* Added scoring for candidate comparison
  - Connect time uses a `exp(-λx)`, where `λ` is `1/overall_ema`, giving us a [0,1] range (curve shape below for `overall_ema` of `50`, values from [0,250] for illustration).
  - Success is the plain EMA, so is already [0,1]
  - Graphsync VerifiedDeal is a `1`, with a weight multiplier of `3`
  - Graphsync FastRetrieval is a `1`, with a weight multiplier of `2` (important, but Verified beats it)
* Candidate comparison involves a dice-roll, line up two candidates, scale their combined values to the `1`, roll a `[0,1]` dice and if falls within the first candidate then it wins.
  - Happily, because we're essentially building a sort algorithm and the prioritywaitqueue basically does a per-candidate sort routine, we don't need to line them up to do the picking, we do it on a a:b comparison basis. It means there's more dice rolling involved (one per a:b comparison rather than one after lining them all up), but I believe that averages out over the set.
* Session config has α and weight options. There's a `DefaultConfig()` which should be used and some sugar utility methods to modify the config according to needs.
* Added a `WithoutRandomness()` config modifier that sets up a fixed dice roll of `0.5` so the best is always selected. This is useful for tests of course.
* Added a bunch of tests to exercise the scoring and ranking, with various permutations of combined metrics.

![Screenshot 2023-05-24 at 2 27 19 pm](https://github.com/filecoin-project/lassie/assets/495647/ea589d6f-a652-4652-b9d0-4a5095850f3d)

In https://github.com/filecoin-project/lassie/issues/92 we still have ttfb and bandwidth. I think the approach with both of these would be the same as the one with connect time (overall EMA, make a [0,1] curve, plot per-SP EMA on that curve). 

There's also concurrency which we're still keeping a strict lid on. We should talk about if/how to incorporate that in scoring.

hannahhoward commented on pkg/retriever/retriever.go:
```diff
@@ -38,14 +38,18 @@ var (
 )
 
 type Session interface {
+	GetStorageProviderTimeout(storageProviderId peer.ID) time.Duration
```
yay conceptual reorg :)

hannahhoward commented on pkg/session/config.go:
```diff
@@ -23,40 +26,119 @@ type Config struct {
+
+// WithoutRandomness removes the dice roll for choosing the best peer, with this
+// set, it will always choose the peer with the highest score.
+func (cfg Config) WithoutRandomness() *Config {
```
love this tool

hannahhoward commented on pkg/session/config.go:
```diff
@@ -23,40 +26,119 @@ type Config struct {
+	cfg.SuccessAlpha = alpha
+	return &cfg
+}
+
```
Probably want configs to adjust the weights no?

hannahhoward replied:
or I mean helper functions, if this is what we're generally going with

rvagg replied:
I'm open to suggestions here btw, I added these cause I realised that you almost never want a `&Config{...}`, you want a `DefaultConfig()` and then customise that, because other wise weights and parameters are all wrong. So these methods are all just sugar.
It's a bit of a foot-gun allowing a bare instantiation but I don't really want to make the variables private.

hannahhoward commented on pkg/session/state.go:
```diff
@@ -209,85 +205,105 @@ func (spt *SessionState) RemoveFromRetrieval(storageProviderId peer.ID, retrieva
+	// exponential decay of connect time, normalised by overall connect time
+	// if we have no connect time data, use the average, if we have no average
+	// use 1
+	λ := 1 / float64(spt.overallConnectTimeMs.getValue(1))
```
I believe λ here is the rate parameter in an exponential distribution. It might be helpful to provide a link to some references about this in a comment. Especially for folks who want to understand what this symbol means.

hannahhoward commented:
agree on the randomness bugs. I've long thought the architecture of the priority wait queue would eventually run into problems. I think you just need one more coordinating go-routine. Essentially, that's what would have the choice whom to run each time the queue frees up, choosing from all the possibilities. I know you're aiming to avoid it for as long as needed, but maybe now we're here.

hannahhoward reviewed (commented):
Generally, this is awesome and the right direction. If we eventually get to cross protocol choices, we'll probably want to reduce the graphsync weights (+ maybe add some protocol prioritizations).
One amazing thing to get to would be instead of protocol racing, we just have everyone running through the priority wait queue with a parallelization of >1. Then again, I have no idea how that works with not "whole request" logic for protocols like Bitswap. TBD and far away.

However, before we do anything further (and I think we need to focus on TTFB next, followed by bandwidth), I think that we need to:
1. Resolve the decision making process so it's "pick one from many" rather than "pick one of two".
2. Focus on how we record what decisions were made and why, so we can see how they affect performance when deployed.

Anyway though, this seems good and we should probably merge it if we can get the tests reliably passing.

rvagg commented:
Good to go

* Updated PWQ to compare all waiting candidates and run the "best" next.
* Replaced `CompareStorageProviders` with `ChooseNextProvider`

hannahhoward commented on pkg/retriever/prioritywaitqueue/prioritywaitqueue.go:
```diff
@@ -159,18 +159,26 @@ func (pwq *priorityWaitQueue[T]) Wait(waitWith T) func() {
-					pwq.cond.Broadcast()
+					// choose the next runner if necessary
+					pwq.chooseNext()
+					if len(pwq.waiters) > 0 {
```
just curious -- what's with adding the `len(pwq.waiters) > 0` here and above?

hannahhoward commented on pkg/session/state.go:
```diff
@@ -209,85 +206,125 @@ func (spt *SessionState) RemoveFromRetrieval(storageProviderId peer.ID, retrieva
+	// if we have no success data, treat it as fully successful
+	score += spt.config.SuccessWeight * sp.success.getValue(1)
+
+	// DEBUG: fmt.Printf("%s v=%v, f=%v, oconn=%d, conn=%d, success=%f, score=%f\n", string(id), v, f, spt.overallConnectTimeMs.value, sp.connectTimeMs.value, sp.success.value, score)
```
seems like this should be a log.Debug call rather than a commented out fmt.Printf

hannahhoward commented on pkg/session/state.go:
```diff
@@ -209,85 +206,125 @@ func (spt *SessionState) RemoveFromRetrieval(storageProviderId peer.ID, retrieva
+	for ii, p := range peers {
+		ind[ii] = ii
+		gsmd, _ := mda[ii].(*metadata.GraphsyncFilecoinV1)
+		scores[ii] = spt.scoreProvider(p, gsmd)
```
just something to consider: maybe at some point we start squaring these values to reduce randomness. (or perhaps make that it's own adjustable parameter)

rvagg replied:
I'm hoping we can come up with ways to be able to know what to do here without guessing too much; I'm trying modelling for now, but insights and A/B testing live might be better.

hannahhoward reviewed (approved):
I've put a few remaining comments, but IMHO, this is LGTM.
